### ISSUE TITLE
Rule proposal: prefer destructuring defaults to Object.assign/rest

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1538]
I found this pattern and it looked like it could be standardized, but it probably isn't very common.

## Fail

```js
  const { foo, bar } = {
    foo: false,
    bar: 1,
    ...options,
  };
```

## Pass

```js
  const { foo = false, bar = 1} = options;
```

**Comment by B2o5T** [comment]
Found this pattern too, I like your proposal 👍

**Comment by sindresorhus** [comment]
I'm using this pattern too. I find it more readable and it also has the benefit of falling back to the default on `undefined`, which object-spread does not.

**Comment by sindresorhus** [comment]
Another fail:

```js
options = {
	foo: false,
	bar: 1,
	...options,
};
```

**Comment by fregante** [comment]
I think your example is still ok when you're going to pass the options object around, for example like you do in `p-retry`

**Comment by sindresorhus** [comment]
> I think your example is still ok when you're going to pass the options object around, for example like you do in p-retry.

A lot of people use my fail example even though they're not passing around options (old me included), and I would argue that's an anti-pattern and makes code less readable. Not sure if we could statically detect this somehow though.

**Comment by fregante** [comment]
I think it varies. Just like you don't want to import "clear" from `mem`, you might not want to be forced to destructure your "options" object. 

And that specific pattern (p-retry) would have to be excluded every time. I suppose this *might* be an acceptable exception though. 

There are more concerns in the discussion I just linked.

**Comment by sindresorhus** [comment]
Ok. Let's just go for the simple case first as presented in the issue description.
